java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java

```diff
@@ -84,6 +85,19 @@ public JavaType resolveTypeSubstitution(JavaType type, JavaType definition) {
     return typeSubstitutionSolver.applySiteSubstitution(type, definition);
   }
 
+  public JavaType resolveTypeSubstitutionWithDiamondOperator(ParametrizedTypeJavaType type, JavaType definition) {
+    ParametrizedTypeJavaType result = type;
+    if (isParametrizedType(definition)) {
+      TypeSubstitution substitution = TypeSubstitutionSolver.substitutionFromSuperType(type, (ParametrizedTypeJavaType) definition);
+      result = (ParametrizedTypeJavaType) typeSubstitutionSolver.applySubstitution(type, substitution);
+    }
+    return typeSubstitutionSolver.objectSubstitution(result);
+  }
+
+  public JavaType parametrizedTypeWithObject(ParametrizedTypeJavaType type) {
```

Comment:
As discussed this should end up being private to Resolve class and work with erasure rather than Object